P-Tree mutatation operations must hand in ValueReader (if any) to sequenceChunker.Done()

This is reproducible but requires this dataset http://demo.noms.io/ahl::bug2213 and my fuse branch https://github.com/ahl/noms/tree/nomsfs as well as fuse (e.g. osxfuse).

removing `b` from the root of that filesystem causes the following panic:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x20 pc=0xecc52]

goroutine 90 [running]:
panic(0x5ba720, 0xc82000e0d0)
	/opt/local/lib/go/src/runtime/panic.go:481 +0x3e6
github.com/attic-labs/noms/go/types.Ref.TargetValue(0xf34774abb3558a72, 0x9165eca6384a2e92, 0xff3545c6, 0x5, 0xc8281b24c0, 0xc82829d100, 0x0, 0x0, 0x0, 0x0)
	/Users/ahl/gosrc/src/github.com/attic-labs/noms/go/types/ref.go:50 +0x42
github.com/attic-labs/noms/go/types.metaTuple.getChildSequence(0xf34774abb3558a72, 0x9165eca6384a2e92, 0xff3545c6, 0x5, 0xc8281b24c0, 0xc82829d100, 0x1, 0xe8ad00, 0xc82000e3d0, 0x0, ...)
	/Users/ahl/gosrc/src/github.com/attic-labs/noms/go/types/meta_sequence.go:45 +0xb2
github.com/attic-labs/noms/go/types.(*sequenceChunker).Done(0xc82006aa10, 0x0, 0x0, 0x0, 0x0)
	/Users/ahl/gosrc/src/github.com/attic-labs/noms/go/types/sequence_chunker.go:242 +0x360
github.com/attic-labs/noms/go/types.(*sequenceChunker).Done(0xc82006a9a0, 0x0, 0x0, 0x0, 0x0)
	/Users/ahl/gosrc/src/github.com/attic-labs/noms/go/types/sequence_chunker.go:223 +0x150
github.com/attic-labs/noms/go/types.Map.splice(0xe8b980, 0xc8281b2840, 0xc82829d160, 0xc8281d4a60, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/ahl/gosrc/src/github.com/attic-labs/noms/go/types/map.go:177 +0x2b7
github.com/attic-labs/noms/go/types.Map.Remove(0xe8b980, 0xc8281b2840, 0xc82829d160, 0xe8ad00, 0xc828294800, 0x0, 0x0, 0x0)
	/Users/ahl/gosrc/src/github.com/attic-labs/noms/go/types/map.go:162 +0xba
main.(*nomsFS).removeCommon(0xc82010cc40, 0xc82829452b, 0x1, 0x792218, 0x0)
	/Users/ahl/src/noms/samples/go/nomsfs/nomsfs.go:379 +0x391
main.(*nomsFS).Unlink(0xc82010cc40, 0xc82829452b, 0x1, 0xc8200f01c8, 0x1)
	/Users/ahl/src/noms/samples/go/nomsfs/nomsfs.go:355 +0x41
github.com/attic-labs/noms/vendor/github.com/attic-labs/go-fuse/fuse/pathfs.(*pathInode).Unlink(0xc82829e060, 0xc82829452b, 0x1, 0xc8200f01c8, 0xc8200b6ac8)
	/Users/ahl/gosrc/src/github.com/attic-labs/noms/vendor/github.com/attic-labs/go-fuse/fuse/pathfs/pathfs.go:436 +0x13e
github.com/attic-labs/noms/vendor/github.com/attic-labs/go-fuse/fuse/nodefs.(*rawBridge).Unlink(0xc82829e0f0, 0xc8200f01b0, 0xc82829452b, 0x1, 0x0)
	/Users/ahl/gosrc/src/github.com/attic-labs/noms/vendor/github.com/attic-labs/go-fuse/fuse/nodefs/fsops.go:300 +0x89
github.com/attic-labs/noms/vendor/github.com/attic-labs/go-fuse/fuse.doUnlink(0xc8200aa000, 0xc8200f0000)
	/Users/ahl/gosrc/src/github.com/attic-labs/noms/vendor/github.com/attic-labs/go-fuse/fuse/opcode.go:295 +0x86
github.com/attic-labs/noms/vendor/github.com/attic-labs/go-fuse/fuse.(*Server).handleRequest(0xc8200aa000, 0xc8200f0000, 0xc8200aa0c8)
	/Users/ahl/gosrc/src/github.com/attic-labs/noms/vendor/github.com/attic-labs/go-fuse/fuse/server.go:394 +0x2ff
created by github.com/attic-labs/noms/vendor/github.com/attic-labs/go-fuse/fuse.(*Server).loop
	/Users/ahl/gosrc/src/github.com/attic-labs/noms/vendor/github.com/attic-labs/go-fuse/fuse/server.go:371 +0xc9
exit status 2
```

The relevant call chain is `Map.Remove()` -> `Map.splice()` -> `sequenceChunker.Done()` and the `vr ValueReader` parameter to `Done()` is `nil`. We crash in the recursive call to `sequenceChunker.Done()` at line 241 since `vr` is `nil`.

--- a/go/types/blob.go
+++ b/go/types/blob.go
@@ -43,7 +43,7 @@ func (b Blob) Splice(idx uint64, deleteCount uint64, data []byte) Blob {
 	d.Chk.True(idx+deleteCount <= b.Len())
 
 	cur := newCursorAtIndex(b.seq, idx)
-	ch := newSequenceChunker(cur, nil, makeBlobLeafChunkFn(b.seq.valueReader()), newIndexedMetaSequenceChunkFn(BlobKind, b.seq.valueReader()), hashValueByte)
+	ch := newSequenceChunker(cur, b.seq.valueReader(), nil, makeBlobLeafChunkFn(b.seq.valueReader()), newIndexedMetaSequenceChunkFn(BlobKind, b.seq.valueReader()), hashValueByte)
 	for deleteCount > 0 {
 		ch.Skip()
 		deleteCount--
@@ -52,7 +52,7 @@ func (b Blob) Splice(idx uint64, deleteCount uint64, data []byte) Blob {
 	for _, v := range data {
 		ch.Append(v)
 	}
-	return newBlob(ch.Done(nil).(indexedSequence))
+	return newBlob(ch.Done().(indexedSequence))
 }
 
 // Collection interface
@@ -177,7 +177,7 @@ func NewBlob(r io.Reader) Blob {
 }
 
 func NewStreamingBlob(r io.Reader, vrw ValueReadWriter) Blob {
-	sc := newEmptySequenceChunker(vrw, makeBlobLeafChunkFn(nil), newIndexedMetaSequenceChunkFn(BlobKind, nil), func(item sequenceItem, rv *rollingValueHasher) {
+	sc := newEmptySequenceChunker(vrw, vrw, makeBlobLeafChunkFn(nil), newIndexedMetaSequenceChunkFn(BlobKind, nil), func(item sequenceItem, rv *rollingValueHasher) {
 		rv.HashByte(item.(byte))
 	})
 
@@ -247,5 +247,5 @@ func NewStreamingBlob(r io.Reader, vrw ValueReadWriter) Blob {
 		sc.parent.Append(b.(metaTuple))
 	}
 
-	return newBlob(sc.Done(vrw).(indexedSequence))
+	return newBlob(sc.Done().(indexedSequence))
 }

--- a/go/types/list.go
+++ b/go/types/list.go
@@ -20,22 +20,22 @@ func newList(seq indexedSequence) List {
 
 // NewList creates a new List where the type is computed from the elements in the list, populated with values, chunking if and when needed.
 func NewList(values ...Value) List {
-	seq := newEmptySequenceChunker(nil, makeListLeafChunkFn(nil), newIndexedMetaSequenceChunkFn(ListKind, nil), hashValueBytes)
+	seq := newEmptySequenceChunker(nil, nil, makeListLeafChunkFn(nil), newIndexedMetaSequenceChunkFn(ListKind, nil), hashValueBytes)
 	for _, v := range values {
 		seq.Append(v)
 	}
-	return newList(seq.Done(nil).(indexedSequence))
+	return newList(seq.Done().(indexedSequence))
 }
 
 // NewStreamingList creates a new List with type t, populated with values, chunking if and when needed. As chunks are created, they're written to vrw -- including the root chunk of the list. Once the caller has closed values, she can read the completed List from the returned channel.
 func NewStreamingList(vrw ValueReadWriter, values <-chan Value) <-chan List {
 	out := make(chan List)
 	go func() {
-		seq := newEmptySequenceChunker(vrw, makeListLeafChunkFn(vrw), newIndexedMetaSequenceChunkFn(ListKind, vrw), hashValueBytes)
+		seq := newEmptySequenceChunker(vrw, vrw, makeListLeafChunkFn(vrw), newIndexedMetaSequenceChunkFn(ListKind, vrw), hashValueBytes)
 		for v := range values {
 			seq.Append(v)
 		}
-		out <- newList(seq.Done(vrw).(indexedSequence))
+		out <- newList(seq.Done().(indexedSequence))
 		close(out)
 	}()
 	return out
@@ -134,7 +134,7 @@ func (l List) Splice(idx uint64, deleteCount uint64, vs ...Value) List {
 	d.Chk.True(idx+deleteCount <= l.Len())
 
 	cur := newCursorAtIndex(l.seq, idx)
-	ch := newSequenceChunker(cur, nil, makeListLeafChunkFn(l.seq.valueReader()), newIndexedMetaSequenceChunkFn(ListKind, l.seq.valueReader()), hashValueBytes)
+	ch := newSequenceChunker(cur, l.seq.valueReader(), nil, makeListLeafChunkFn(l.seq.valueReader()), newIndexedMetaSequenceChunkFn(ListKind, l.seq.valueReader()), hashValueBytes)
 	for deleteCount > 0 {
 		ch.Skip()
 		deleteCount--
@@ -143,7 +143,7 @@ func (l List) Splice(idx uint64, deleteCount uint64, vs ...Value) List {
 	for _, v := range vs {
 		ch.Append(v)
 	}
-	return newList(ch.Done(nil).(indexedSequence))
+	return newList(ch.Done().(indexedSequence))
 }
 
 func (l List) Insert(idx uint64, vs ...Value) List {

--- a/go/types/list_test.go
+++ b/go/types/list_test.go
@@ -1047,8 +1047,8 @@ func TestListRemoveLastWhenNotLoaded(t *testing.T) {
 	defer normalProductionChunks()
 
 	vs := NewTestValueStore()
-	reload := func(v Value) Value {
-		return vs.ReadValue(vs.WriteValue(v).TargetHash())
+	reload := func(l List) List {
+		return vs.ReadValue(vs.WriteValue(l).TargetHash()).(List)
 	}
 
 	tl := newTestList(1024)
@@ -1056,8 +1056,7 @@ func TestListRemoveLastWhenNotLoaded(t *testing.T) {
 
 	for len(tl) > 0 {
 		tl = tl[:len(tl)-1]
-		nl = nl.RemoveAt(uint64(len(tl)))
-		nl = reload(nl).(List)
+		nl = reload(nl.RemoveAt(uint64(len(tl))))
 		assert.True(tl.toList().Equals(nl))
 	}
 }

--- a/go/types/list_test.go
+++ b/go/types/list_test.go
@@ -1039,3 +1039,25 @@ func TestListTypeAfterMutations(t *testing.T) {
 	test(15, listLeafSequence{})
 	test(1500, indexedMetaSequence{})
 }
+
+func TestListRemoveLastWhenNotLoaded(t *testing.T) {
+	assert := assert.New(t)
+
+	smallTestChunks()
+	defer normalProductionChunks()
+
+	vs := NewTestValueStore()
+	reload := func(v Value) Value {
+		return vs.ReadValue(vs.WriteValue(v).TargetHash())
+	}
+
+	tl := newTestList(1024)
+	nl := tl.toList()
+
+	for len(tl) > 0 {
+		tl = tl[:len(tl)-1]
+		nl = nl.RemoveAt(uint64(len(tl)))
+		nl = reload(nl).(List)
+		assert.True(tl.toList().Equals(nl))
+	}
+}

--- a/go/types/map.go
+++ b/go/types/map.go
@@ -28,13 +28,13 @@ func mapHashValueBytes(item sequenceItem, rv *rollingValueHasher) {
 
 func NewMap(kv ...Value) Map {
 	entries := buildMapData(kv)
-	seq := newEmptySequenceChunker(nil, makeMapLeafChunkFn(nil), newOrderedMetaSequenceChunkFn(MapKind, nil), mapHashValueBytes)
+	seq := newEmptySequenceChunker(nil, nil, makeMapLeafChunkFn(nil), newOrderedMetaSequenceChunkFn(MapKind, nil), mapHashValueBytes)
 
 	for _, entry := range entries {
 		seq.Append(entry)
 	}
 
-	return newMap(seq.Done(nil).(orderedSequence))
+	return newMap(seq.Done().(orderedSequence))
 }
 
 func NewStreamingMap(vrw ValueReadWriter, kvs <-chan Value) <-chan Map {
@@ -174,7 +174,7 @@ func (m Map) Remove(k Value) Map {
 }
 
 func (m Map) splice(cur *sequenceCursor, deleteCount uint64, vs ...mapEntry) Map {
-	ch := newSequenceChunker(cur, nil, makeMapLeafChunkFn(m.seq.valueReader()), newOrderedMetaSequenceChunkFn(MapKind, m.seq.valueReader()), mapHashValueBytes)
+	ch := newSequenceChunker(cur, m.seq.valueReader(), nil, makeMapLeafChunkFn(m.seq.valueReader()), newOrderedMetaSequenceChunkFn(MapKind, m.seq.valueReader()), mapHashValueBytes)
 	for deleteCount > 0 {
 		ch.Skip()
 		deleteCount--
@@ -183,7 +183,7 @@ func (m Map) splice(cur *sequenceCursor, deleteCount uint64, vs ...mapEntry) Map
 	for _, v := range vs {
 		ch.Append(v)
 	}
-	return newMap(ch.Done(nil).(orderedSequence))
+	return newMap(ch.Done().(orderedSequence))
 }
 
 func (m Map) getCursorAtValue(v Value) (cur *sequenceCursor, found bool) {

--- a/go/types/map_test.go
+++ b/go/types/map_test.go
@@ -1131,3 +1131,28 @@ func TestCompoundMapWithValuesOfEveryType(t *testing.T) {
 		assert.Equal(len(kvs)/2, int(m.Len()))
 	}
 }
+
+func TestMapRemoveLastWhenNotLoaded(t *testing.T) {
+	assert := assert.New(t)
+
+	smallTestChunks()
+	defer normalProductionChunks()
+
+	vs := NewTestValueStore()
+	reload := func(v Value) Value {
+		return vs.ReadValue(vs.WriteValue(v).TargetHash())
+	}
+
+	tm := getTestNativeOrderMap(4)
+	nm := tm.toMap()
+
+	for len(tm.entries) > 0 {
+		entr := tm.entries
+		last := entr[len(entr)-1]
+		entr = entr[:len(entr)-1]
+		tm.entries = entr
+		nm = nm.Remove(last.key)
+		nm = reload(nm).(Map)
+		assert.True(tm.toMap().Equals(nm))
+	}
+}

--- a/go/types/map_test.go
+++ b/go/types/map_test.go
@@ -1139,8 +1139,8 @@ func TestMapRemoveLastWhenNotLoaded(t *testing.T) {
 	defer normalProductionChunks()
 
 	vs := NewTestValueStore()
-	reload := func(v Value) Value {
-		return vs.ReadValue(vs.WriteValue(v).TargetHash())
+	reload := func(m Map) Map {
+		return vs.ReadValue(vs.WriteValue(m).TargetHash()).(Map)
 	}
 
 	tm := getTestNativeOrderMap(4)
@@ -1151,8 +1151,7 @@ func TestMapRemoveLastWhenNotLoaded(t *testing.T) {
 		last := entr[len(entr)-1]
 		entr = entr[:len(entr)-1]
 		tm.entries = entr
-		nm = nm.Remove(last.key)
-		nm = reload(nm).(Map)
+		nm = reload(nm.Remove(last.key))
 		assert.True(tm.toMap().Equals(nm))
 	}
 }

--- a/go/types/mutator.go
+++ b/go/types/mutator.go
@@ -28,7 +28,7 @@ func (mx *mapMutator) Finish() Map {
 		mx.oc = nil
 	}()
 
-	seq := newEmptySequenceChunker(mx.vrw, makeMapLeafChunkFn(mx.vrw), newOrderedMetaSequenceChunkFn(MapKind, mx.vrw), mapHashValueBytes)
+	seq := newEmptySequenceChunker(mx.vrw, mx.vrw, makeMapLeafChunkFn(mx.vrw), newOrderedMetaSequenceChunkFn(MapKind, mx.vrw), mapHashValueBytes)
 
 	// I tried splitting this up so that the iteration ran in a separate goroutine from the Append'ing, but it actually made things a bit slower when I ran a test.
 	iter := mx.oc.NewIterator()
@@ -36,5 +36,5 @@ func (mx *mapMutator) Finish() Map {
 	for iter.Next() {
 		seq.Append(iter.Op())
 	}
-	return newMap(seq.Done(mx.vrw).(orderedSequence))
+	return newMap(seq.Done().(orderedSequence))
 }

--- a/go/types/sequence_chunker.go
+++ b/go/types/sequence_chunker.go
@@ -10,6 +10,7 @@ type hashValueBytesFn func(item sequenceItem, rv *rollingValueHasher)
 
 type sequenceChunker struct {
 	cur                        *sequenceCursor
+	vr                         ValueReader
 	vw                         ValueWriter
 	parent                     *sequenceChunker
 	current                    []sequenceItem
@@ -23,11 +24,11 @@ type sequenceChunker struct {
 // makeChunkFn takes a sequence of items to chunk, and returns the result of chunking those items, a tuple of a reference to that chunk which can itself be chunked + its underlying value.
 type makeChunkFn func(values []sequenceItem) (Collection, orderedKey, uint64)
 
-func newEmptySequenceChunker(vw ValueWriter, makeChunk, parentMakeChunk makeChunkFn, hashValueBytes hashValueBytesFn) *sequenceChunker {
-	return newSequenceChunker(nil, vw, makeChunk, parentMakeChunk, hashValueBytes)
+func newEmptySequenceChunker(vr ValueReader, vw ValueWriter, makeChunk, parentMakeChunk makeChunkFn, hashValueBytes hashValueBytesFn) *sequenceChunker {
+	return newSequenceChunker(nil, vr, vw, makeChunk, parentMakeChunk, hashValueBytes)
 }
 
-func newSequenceChunker(cur *sequenceCursor, vw ValueWriter, makeChunk, parentMakeChunk makeChunkFn, hashValueBytes hashValueBytesFn) *sequenceChunker {
+func newSequenceChunker(cur *sequenceCursor, vr ValueReader, vw ValueWriter, makeChunk, parentMakeChunk makeChunkFn, hashValueBytes hashValueBytesFn) *sequenceChunker {
 	// |cur| will be nil if this is a new sequence, implying this is a new tree, or the tree has grown in height relative to its original chunked form.
 	d.Chk.True(makeChunk != nil)
 	d.Chk.True(parentMakeChunk != nil)
@@ -35,6 +36,7 @@ func newSequenceChunker(cur *sequenceCursor, vw ValueWriter, makeChunk, parentMa
 
 	sc := &sequenceChunker{
 		cur,
+		vr,
 		vw,
 		nil,
 		[]sequenceItem{},
@@ -156,7 +158,7 @@ func (sc *sequenceChunker) createParent() {
 		// Clone the parent cursor because otherwise calling cur.advance() will affect our parent - and vice versa - in surprising ways. Instead, Skip moves forward our parent's cursor if we advance across a boundary.
 		parent = sc.cur.parent.clone()
 	}
-	sc.parent = newSequenceChunker(parent, sc.vw, sc.parentMakeChunk, sc.parentMakeChunk, metaHashValueBytes)
+	sc.parent = newSequenceChunker(parent, sc.vr, sc.vw, sc.parentMakeChunk, sc.parentMakeChunk, metaHashValueBytes)
 	sc.parent.isLeaf = false
 }
 
@@ -201,8 +203,7 @@ func (sc *sequenceChunker) anyPending() bool {
 }
 
 // Returns the root sequence of the resulting tree. The logic here is subtle, but hopefully correct and understandable. See comments inline.
-func (sc *sequenceChunker) Done(vr ValueReader) sequence {
-	d.Chk.True((vr == nil) == (sc.vw == nil))
+func (sc *sequenceChunker) Done() sequence {
 	d.Chk.False(sc.done)
 	sc.done = true
 
@@ -217,7 +218,7 @@ func (sc *sequenceChunker) Done(vr ValueReader) sequence {
 			sc.handleChunkBoundary()
 		}
 
-		return sc.parent.Done(vr)
+		return sc.parent.Done()
 	}
 
 	// At this point, we know this chunker contains, in |current| every item at this level of the resulting tree. To see this, consider that there are two ways a chunker can enter items into its |current|: (1) as the result of resume() with the cursor on anything other than the first item in the sequence, and (2) as a result of a child chunker hitting an explicit chunk boundary during either Append() or finalize(). The only way there can be no items in some parent chunker's |current| is if this chunker began with cursor within its first existing chunk (and thus all parents resume()'d with a cursor on their first item) and continued through all sebsequent items without creating any explicit chunk boundaries (and thus never sent any items up to a parent as a result of chunking). Therefore, this chunker's current must contain all items within the current sequence.
@@ -235,7 +236,7 @@ func (sc *sequenceChunker) Done(vr ValueReader) sequence {
 	mt := sc.current[0].(metaTuple)
 
 	for {
-		child := mt.getChildSequence(vr)
+		child := mt.getChildSequence(sc.vr)
 		if _, ok := child.(metaSequence); !ok || child.seqLen() > 1 {
 			return child
 		}

--- a/go/types/set.go
+++ b/go/types/set.go
@@ -21,13 +21,13 @@ func newSet(seq orderedSequence) Set {
 
 func NewSet(v ...Value) Set {
 	data := buildSetData(v)
-	seq := newEmptySequenceChunker(nil, makeSetLeafChunkFn(nil), newOrderedMetaSequenceChunkFn(SetKind, nil), hashValueBytes)
+	seq := newEmptySequenceChunker(nil, nil, makeSetLeafChunkFn(nil), newOrderedMetaSequenceChunkFn(SetKind, nil), hashValueBytes)
 
 	for _, v := range data {
 		seq.Append(v)
 	}
 
-	return newSet(seq.Done(nil).(orderedSequence))
+	return newSet(seq.Done().(orderedSequence))
 }
 
 // Computes the diff from |last| to |s| using "best" algorithm, which balances returning results early vs completing quickly.
@@ -138,7 +138,7 @@ func (s Set) Remove(values ...Value) Set {
 }
 
 func (s Set) splice(cur *sequenceCursor, deleteCount uint64, vs ...Value) Set {
-	ch := newSequenceChunker(cur, nil, makeSetLeafChunkFn(s.seq.valueReader()), newOrderedMetaSequenceChunkFn(SetKind, s.seq.valueReader()), hashValueBytes)
+	ch := newSequenceChunker(cur, s.seq.valueReader(), nil, makeSetLeafChunkFn(s.seq.valueReader()), newOrderedMetaSequenceChunkFn(SetKind, s.seq.valueReader()), hashValueBytes)
 	for deleteCount > 0 {
 		ch.Skip()
 		deleteCount--
@@ -148,7 +148,7 @@ func (s Set) splice(cur *sequenceCursor, deleteCount uint64, vs ...Value) Set {
 		ch.Append(v)
 	}
 
-	ns := newSet(ch.Done(nil).(orderedSequence))
+	ns := newSet(ch.Done().(orderedSequence))
 	return ns
 }
 

--- a/go/types/set_test.go
+++ b/go/types/set_test.go
@@ -926,3 +926,26 @@ func TestChunkedSetWithValuesOfEveryType(t *testing.T) {
 		assert.Equal(len(vs), int(s.Len()))
 	}
 }
+
+func TestSetRemoveLastWhenNotLoaded(t *testing.T) {
+	assert := assert.New(t)
+
+	smallTestChunks()
+	defer normalProductionChunks()
+
+	vs := NewTestValueStore()
+	reload := func(v Value) Value {
+		return vs.ReadValue(vs.WriteValue(v).TargetHash())
+	}
+
+	ts := getTestNativeOrderSet(8)
+	ns := ts.toSet()
+
+	for len(ts) > 0 {
+		last := ts[len(ts)-1]
+		ts = ts[:len(ts)-1]
+		ns = ns.Remove(last)
+		ns = reload(ns).(Set)
+		assert.True(ts.toSet().Equals(ns))
+	}
+}

--- a/go/types/set_test.go
+++ b/go/types/set_test.go
@@ -934,8 +934,8 @@ func TestSetRemoveLastWhenNotLoaded(t *testing.T) {
 	defer normalProductionChunks()
 
 	vs := NewTestValueStore()
-	reload := func(v Value) Value {
-		return vs.ReadValue(vs.WriteValue(v).TargetHash())
+	reload := func(s Set) Set {
+		return vs.ReadValue(vs.WriteValue(s).TargetHash()).(Set)
 	}
 
 	ts := getTestNativeOrderSet(8)
@@ -944,8 +944,7 @@ func TestSetRemoveLastWhenNotLoaded(t *testing.T) {
 	for len(ts) > 0 {
 		last := ts[len(ts)-1]
 		ts = ts[:len(ts)-1]
-		ns = ns.Remove(last)
-		ns = reload(ns).(Set)
+		ns = reload(ns.Remove(last))
 		assert.True(ts.toSet().Equals(ns))
 	}
 }

--- a/js/src/blob.js
+++ b/js/src/blob.js
@@ -7,8 +7,7 @@
 import * as Bytes from './bytes.js';
 import Collection from './collection.js';
 import RollingValueHasher from './rolling-value-hasher.js';
-import SequenceChunker from './sequence-chunker.js';
-import {chunkSequence} from './sequence-chunker.js';
+import {default as SequenceChunker, chunkSequence} from './sequence-chunker.js';
 import type {EqualsFn} from './edit-distance.js';
 import type {ValueReader, ValueReadWriter} from './value-store.js';
 import type {makeChunkFn} from './sequence-chunker.js';
@@ -22,7 +21,7 @@ import {hashValueByte} from './rolling-value-hasher.js';
 
 export default class Blob extends Collection<IndexedSequence> {
   constructor(bytes: Uint8Array) {
-    const chunker = new SequenceChunker(null, null, newBlobLeafChunkFn(null),
+    const chunker = new SequenceChunker(null, null, null, newBlobLeafChunkFn(null),
         newIndexedMetaSequenceChunkFn(Kind.Blob, null), blobHashValueBytes);
 
     for (let i = 0; i < bytes.length; i++) {
@@ -45,7 +44,7 @@ export default class Blob extends Collection<IndexedSequence> {
   splice(idx: number, deleteCount: number, insert: Uint8Array): Promise<Blob> {
     const vr = this.sequence.vr;
     return this.sequence.newCursorAt(idx).then(cursor =>
-      chunkSequence(cursor, Array.from(insert), deleteCount, newBlobLeafChunkFn(vr),
+      chunkSequence(cursor, vr, Array.from(insert), deleteCount, newBlobLeafChunkFn(vr),
                     newIndexedMetaSequenceChunkFn(Kind.Blob, vr, null),
                     hashValueByte)).then(s => Blob.fromSequence(s));
   }
@@ -180,7 +179,7 @@ export class BlobWriter {
 
   constructor(vrw: ?ValueReadWriter) {
     this._state = 'writable';
-    this._chunker = new SequenceChunker(null, vrw, newBlobLeafChunkFn(vrw),
+    this._chunker = new SequenceChunker(null, vrw, vrw, newBlobLeafChunkFn(vrw),
         newIndexedMetaSequenceChunkFn(Kind.Blob, vrw), blobHashValueBytes);
     this._vrw = vrw;
   }

--- a/js/src/list-test.js
+++ b/js/src/list-test.js
@@ -15,6 +15,7 @@ import {equals} from './compare.js';
 import {invariant, notNull} from './assert.js';
 import {newStruct} from './struct.js';
 import type Value from './value.js';
+import {smallTestChunks, normalProductionChunks} from './rolling-value-hasher.js';
 import {
   makeListType,
   makeRefType,
@@ -591,10 +592,14 @@ suite('ListWriter', () => {
   let db;
 
   setup(() => {
+    smallTestChunks();
     db = new TestDatabase();
   });
 
-  teardown((): Promise<void> => db.close());
+  teardown(async () => {
+    normalProductionChunks();
+    await db.close();
+  });
 
   test('ListWriter', async () => {
     const values = intSequence(15);
@@ -695,5 +700,22 @@ suite('ListWriter', () => {
 
     await t(15, ListLeafSequence);
     await t(1500, IndexedMetaSequence);
+  });
+
+  test('Remove last when not loaded', async () => {
+    const reload = async (l: List): Promise<List> => {
+      const l2 = await db.readValue(db.writeValue(l).targetHash);
+      invariant(l2 instanceof List);
+      return l2;
+    };
+
+    let vals = intSequence(64);
+    let list = new List(vals);
+
+    while (vals.length > 0) {
+      vals = vals.slice(0, vals.length - 1);
+      list = await list.remove(vals.length).then(reload);
+      assert.isTrue(equals(new List(vals), list));
+    }
   });
 });

--- a/js/src/list-test.js
+++ b/js/src/list-test.js
@@ -298,10 +298,15 @@ suite('CompoundList', () => {
   let db;
 
   setup(() => {
+    smallTestChunks();
     db = new TestDatabase();
   });
 
-  teardown((): Promise<void> => db.close());
+  teardown(async () => {
+    normalProductionChunks();
+    await db.close();
+  });
+
   function build(): List {
     const l1 = new List(['a', 'b']);
     const r1 = db.writeValue(l1);
@@ -364,6 +369,23 @@ suite('CompoundList', () => {
     assert.deepEqual(
         [{done: false, value: 'a'}, {done: false, value: 'b'}, {done: true}, {done: true}],
         values);
+  });
+
+  test('Remove last when not loaded', async () => {
+    const reload = async (l: List): Promise<List> => {
+      const l2 = await db.readValue(db.writeValue(l).targetHash);
+      invariant(l2 instanceof List);
+      return l2;
+    };
+
+    let vals = intSequence(64);
+    let list = new List(vals);
+
+    while (vals.length > 0) {
+      vals = vals.slice(0, vals.length - 1);
+      list = await list.remove(vals.length).then(reload);
+      assert.isTrue(equals(new List(vals), list));
+    }
   });
 });
 
@@ -592,14 +614,10 @@ suite('ListWriter', () => {
   let db;
 
   setup(() => {
-    smallTestChunks();
     db = new TestDatabase();
   });
 
-  teardown(async () => {
-    normalProductionChunks();
-    await db.close();
-  });
+  teardown((): Promise<void> => db.close());
 
   test('ListWriter', async () => {
     const values = intSequence(15);
@@ -700,22 +718,5 @@ suite('ListWriter', () => {
 
     await t(15, ListLeafSequence);
     await t(1500, IndexedMetaSequence);
-  });
-
-  test('Remove last when not loaded', async () => {
-    const reload = async (l: List): Promise<List> => {
-      const l2 = await db.readValue(db.writeValue(l).targetHash);
-      invariant(l2 instanceof List);
-      return l2;
-    };
-
-    let vals = intSequence(64);
-    let list = new List(vals);
-
-    while (vals.length > 0) {
-      vals = vals.slice(0, vals.length - 1);
-      list = await list.remove(vals.length).then(reload);
-      assert.isTrue(equals(new List(vals), list));
-    }
   });
 });

--- a/js/src/list.js
+++ b/js/src/list.js
@@ -9,7 +9,7 @@ import type {ValueReader, ValueReadWriter} from './value-store.js';
 import type {Splice} from './edit-distance.js';
 import type Value from './value.js'; // eslint-disable-line no-unused-vars
 import type {AsyncIterator} from './async-iterator.js';
-import {chunkSequence, chunkSequenceSync} from './sequence-chunker.js';
+import {default as SequenceChunker, chunkSequence, chunkSequenceSync} from './sequence-chunker.js';
 import Collection from './collection.js';
 import {IndexedSequence, IndexedSequenceIterator} from './indexed-sequence.js';
 import {diff} from './indexed-sequence-diff.js';
@@ -23,7 +23,6 @@ import {getValueChunks} from './sequence.js';
 import {makeListType, makeUnionType, getTypeOfValue} from './type.js';
 import {equals} from './compare.js';
 import {Kind} from './noms-kind.js';
-import SequenceChunker from './sequence-chunker.js';
 import {DEFAULT_MAX_SPLICE_MATRIX_SIZE} from './edit-distance.js';
 import {hashValueBytes} from './rolling-value-hasher.js';
 
@@ -55,7 +54,7 @@ export default class List<T: Value> extends Collection<IndexedSequence> {
   splice(idx: number, deleteCount: number, ...insert: Array<T>): Promise<List<T>> {
     const vr = this.sequence.vr;
     return this.sequence.newCursorAt(idx).then(cursor =>
-      chunkSequence(cursor, insert, deleteCount, newListLeafChunkFn(vr),
+      chunkSequence(cursor, vr, insert, deleteCount, newListLeafChunkFn(vr),
                     newIndexedMetaSequenceChunkFn(Kind.List, vr, null),
                     hashValueBytes)).then(s => List.fromSequence(s));
   }
@@ -163,7 +162,7 @@ export class ListWriter<T: Value> {
 
   constructor(vrw: ?ValueReadWriter) {
     this._state = 'writable';
-    this._chunker = new SequenceChunker(null, vrw, newListLeafChunkFn(vrw),
+    this._chunker = new SequenceChunker(null, vrw, vrw, newListLeafChunkFn(vrw),
         newIndexedMetaSequenceChunkFn(Kind.List, vrw, vrw), hashValueBytes);
     this._vrw = vrw;
   }

--- a/js/src/map-test.js
+++ b/js/src/map-test.js
@@ -31,6 +31,7 @@ import {OrderedKey, MetaTuple, newMapMetaSequence} from './meta-sequence.js';
 import type {ValueReadWriter} from './value-store.js';
 import {compare, equals} from './compare.js';
 import {OrderedMetaSequence} from './meta-sequence.js';
+import {smallTestChunks, normalProductionChunks} from './rolling-value-hasher.js';
 import {
   makeMapType,
   makeRefType,
@@ -441,10 +442,14 @@ suite('CompoundMap', () => {
   let db;
 
   setup(() => {
+    smallTestChunks();
     db = new TestDatabase();
   });
 
-  teardown((): Promise<void> => db.close());
+  teardown(async () => {
+    normalProductionChunks();
+    await db.close();
+  });
 
   function build(vwr: ValueReadWriter): Array<Map> {
     const l1 = new Map([['a', false], ['b', false]]);
@@ -725,7 +730,14 @@ suite('CompoundMap', () => {
       assert.isTrue(equals(m.type, makeMapType(numberType, numberType)));
     }
 
-    await t(10, MapLeafSequence);
+    try {
+      // This test shouldn't rely on a specific chunk size producing a leaf sequence, but it does.
+      normalProductionChunks();
+      await t(10, MapLeafSequence);
+    } finally {
+      smallTestChunks();
+    }
+
     await t(1000, OrderedMetaSequence);
   });
 
@@ -774,6 +786,24 @@ suite('CompoundMap', () => {
       m = await m.delete(kv[0]);
       assert.isFalse(await m.has(kv[0]));
       assert.strictEqual(kvs.length, m.size);
+    }
+  });
+
+  test('Remove last when not loaded', async () => {
+    const reload = async (m: Map): Promise<Map> => {
+      const m2 = await db.readValue(db.writeValue(m).targetHash);
+      invariant(m2 instanceof Map);
+      return m2;
+    };
+
+    let kvs = intSequence(64).map(n => [n, n]);
+    let map = new Map(kvs);
+
+    while (kvs.length > 0) {
+      const lastKey = kvs[kvs.length - 1][0];
+      kvs = kvs.slice(0, kvs.length - 1);
+      map = await map.delete(lastKey).then(reload);
+      assert.isTrue(equals(new Map(kvs), map));
     }
   });
 });

--- a/js/src/map.js
+++ b/js/src/map.js
@@ -142,7 +142,7 @@ export default class Map<K: Value, V: Value> extends
   _splice(cursor: OrderedSequenceCursor, insert: Array<MapEntry<K, V>>, remove: number):
       Promise<Map<K, V>> {
     const vr = this.sequence.vr;
-    return chunkSequence(cursor, insert, remove, newMapLeafChunkFn(vr),
+    return chunkSequence(cursor, vr, insert, remove, newMapLeafChunkFn(vr),
                          newOrderedMetaSequenceChunkFn(Kind.Map, vr),
                          mapHashValueBytes).then(s => Map.fromSequence(s));
   }

--- a/js/src/sequence-chunker.js
+++ b/js/src/sequence-chunker.js
@@ -20,13 +20,14 @@ export type hashValueBytesFn<T> = (item: T, rv: RollingValueHasher) => void;
 
 export async function chunkSequence<T, S: Sequence<T>>(
     cursor: SequenceCursor,
+    vr: ?ValueReader,
     insert: Array<T>,
     remove: number,
     makeChunk: makeChunkFn<T, S>,
     parentMakeChunk: makeChunkFn<MetaTuple, MetaSequence>,
     hashValueBytes: hashValueBytesFn): Promise<Sequence> {
 
-  const chunker = new SequenceChunker(cursor, null, makeChunk, parentMakeChunk, hashValueBytes);
+  const chunker = new SequenceChunker(cursor, vr, null, makeChunk, parentMakeChunk, hashValueBytes);
   if (cursor) {
     await chunker.resume();
   }
@@ -40,7 +41,7 @@ export async function chunkSequence<T, S: Sequence<T>>(
 
   insert.forEach(i => chunker.append(i));
 
-  return chunker.done(null);
+  return chunker.done();
 }
 
 // Like |chunkSequence|, but without an existing cursor (implying this is a new collection), so it
@@ -52,7 +53,7 @@ export function chunkSequenceSync<T, S: Sequence<T>>(
     parentMakeChunk: makeChunkFn<MetaTuple, MetaSequence>,
     hashValueBytes: hashValueBytesFn): Sequence {
 
-  const chunker = new SequenceChunker(null, null, makeChunk, parentMakeChunk, hashValueBytes);
+  const chunker = new SequenceChunker(null, null, null, makeChunk, parentMakeChunk, hashValueBytes);
 
   insert.forEach(i => chunker.append(i));
 
@@ -61,6 +62,7 @@ export function chunkSequenceSync<T, S: Sequence<T>>(
 
 export default class SequenceChunker<T, S: Sequence<T>> {
   _cursor: ?SequenceCursor<T, S>;
+  _vr: ?ValueReader;
   _vw: ?ValueWriter;
   _parent: ?SequenceChunker<MetaTuple, MetaSequence>;
   _current: Array<T>;
@@ -71,9 +73,10 @@ export default class SequenceChunker<T, S: Sequence<T>> {
   _rv: RollingValueHasher;
   _done: boolean;
 
-  constructor(cursor: ?SequenceCursor, vw: ?ValueWriter, makeChunk: makeChunkFn,
+  constructor(cursor: ?SequenceCursor, vr: ?ValueReader, vw: ?ValueWriter, makeChunk: makeChunkFn,
               parentMakeChunk: makeChunkFn, hashValueBytes: hashValueBytesFn) {
     this._cursor = cursor;
+    this._vr = vr;
     this._vw = vw;
     this._parent = null;
     this._current = [];
@@ -193,6 +196,7 @@ export default class SequenceChunker<T, S: Sequence<T>> {
     invariant(!this._parent);
     this._parent = new SequenceChunker(
         this._cursor && this._cursor.parent ? this._cursor.parent.clone() : null,
+        this._vr,
         this._vw,
         this._parentMakeChunk,
         this._parentMakeChunk,
@@ -242,8 +246,7 @@ export default class SequenceChunker<T, S: Sequence<T>> {
 
   // Returns the root sequence of the resulting tree. The logic here is subtle, but hopefully
   // correct and understandable. See comments inline.
-  async done(vr: ?ValueReader): Promise<Sequence> {
-    invariant(!vr === !this._vw);
+  async done(): Promise<Sequence> {
     invariant(!this._done);
     this._done = true;
 
@@ -261,7 +264,7 @@ export default class SequenceChunker<T, S: Sequence<T>> {
         this.handleChunkBoundary();
       }
 
-      return notNull(this._parent).done(vr);
+      return notNull(this._parent).done();
     }
 
     // At this point, we know this chunker contains, in |current| every item at this level of the
@@ -292,7 +295,7 @@ export default class SequenceChunker<T, S: Sequence<T>> {
     invariant(!this._isLeaf && this._current.length === 1);
     const mt = this._current[0];
     invariant(mt instanceof MetaTuple);
-    let seq = await mt.getChildSequence(vr);
+    let seq = await mt.getChildSequence(this._vr);
 
     while (seq.isMeta && seq.length === 1) {
       seq = await seq.getChildSequence(0);

--- a/js/src/set-test.js
+++ b/js/src/set-test.js
@@ -28,6 +28,7 @@ import type Value from './value.js';
 import {invariant, notNull} from './assert.js';
 import {newStruct} from './struct.js';
 import {OrderedMetaSequence} from './meta-sequence.js';
+import {smallTestChunks, normalProductionChunks} from './rolling-value-hasher.js';
 import {
   makeRefType,
   makeSetType,
@@ -356,10 +357,14 @@ suite('CompoundSet', () => {
   let db;
 
   setup(() => {
+    smallTestChunks();
     db = new TestDatabase();
   });
 
-  teardown((): Promise<void> => db.close());
+  teardown(async () => {
+    normalProductionChunks();
+    await db.close();
+  });
 
   function build(vwr: ValueReadWriter, values: Array<string>): Set {
     assert.isTrue(values.length > 1 && Math.log2(values.length) % 1 === 0);
@@ -646,5 +651,23 @@ suite('CompoundSet', () => {
 
     await t(10, SetLeafSequence);
     await t(2000, OrderedMetaSequence);
+  });
+
+  test('Remove last when not loaded', async () => {
+    const reload = async (s: Set): Promise<Set> => {
+      const s2 = await db.readValue(db.writeValue(s).targetHash);
+      invariant(s2 instanceof Set);
+      return s2;
+    };
+
+    let vals = intSequence(64);
+    let set = new Set(vals);
+
+    while (vals.length > 0) {
+      const last = vals[vals.length - 1];
+      vals = vals.slice(0, vals.length - 1);
+      set = await set.delete(last).then(reload);
+      assert.isTrue(equals(new Set(vals), set));
+    }
   });
 });

--- a/js/src/set.js
+++ b/js/src/set.js
@@ -97,7 +97,7 @@ export default class Set<T: Value> extends Collection<OrderedSequence> {
   _splice(cursor: OrderedSequenceCursor, insert: Array<T>, remove: number):
       Promise<Set<T>> {
     const vr = this.sequence.vr;
-    return chunkSequence(cursor, insert, remove, newSetLeafChunkFn(vr),
+    return chunkSequence(cursor, vr, insert, remove, newSetLeafChunkFn(vr),
                          newOrderedMetaSequenceChunkFn(Kind.Set, vr),
                          hashValueBytes).then(s => Set.fromSequence(s));
   }

